### PR TITLE
Feature/wildcard matching

### PR DESCRIPTION
**Pull request by mstrYoda**
**Please provide a description of this PR:**

Add wildcard path matching. This pr makes possible to write authz. paths like:

- "/test/*/resource" 
- "/test/\*/resource/\*/subResource"

Resolves: https://github.com/istio/istio/issues/16585

I've test it locally and it works fine.

Here is the AuthorizationPolicy I wrote:

```
apiVersion: v1
items:
- apiVersion: security.istio.io/v1beta1
  kind: AuthorizationPolicy
  metadata:
    name: test-auth-policy
  spec:
    action: DENY
    rules:
    - to:
      - operation:
          methods:
          - GET
          paths:
          - "/test/*/example"
          - "/test/*/example/*/subpath"
      when:
      - key: request.headers[test]
        notValues:
        - example
    - to:
      - operation:
          methods:
          - POST
          paths:
          - "/test/*/example"
      when:
      - key: request.headers[test]
        notValues:
        - example
    selector:
      matchLabels:
        app: nginx
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

Here is the example requests and expected behaviors:

istio-authz-ss

### CODE REVIEW

**Comment by sschepens** [comment]
@howardjohn we also would need this feature, or be able to specify regexes.

> Additionally, I am concerned about the matching discrepancies with Authz and VirtualService with this. It seems odd I can allow /foo/*/bar but I cannot then route on the same decision.

This also comes to mind, why do we handle Authorization matching in a different way that we do Route matching? i seems as though they should be the same or very similar. Shouldn't we just migrate Authorization matching to use `StringMatch` ?

File: pilot/pkg/security/authz/matcher/string.go [hunk 1]
```diff
@@ -63,6 +63,8 @@ func StringMatcherWithPrefix(v, prefix string) *matcherpb.StringMatcher {
 				Prefix: prefix + strings.TrimSuffix(v, "*"),
 			},
 		}
+	case strings.Contains(v, "*"):
+		return StringMatcherRegex(prefix + strings.ReplaceAll(v, "*", ".+"))
```
[the remark this answers is not in the document]
Comment (mstrYoda) [reply pilot/pkg/security/authz/matcher/string.go:67]: Hi @howardjohn, what is your opinion on my last comment?

**Comment by qRainBirdp** [comment]
Has this function already beeHas this function already been available in the official version of istio?
We really need it！

**Comment by mstrYoda** [comment]
> @yangminzhu Hi, what about securty team? Does anyone sent this PR to tham?
> 
> We really need it 💯☺️🚀🎉



> Has this function already beeHas this function already been available in the official version of istio? We really need it！

Is there any Istio Security Team member here? Can anyone mention someone that can review this feature?

**Comment by craigbox** [comment]
Are you able to drop by a working group meeting to discuss?

See https://github.com/istio/istio/pull/35641#pullrequestreview-783642732:

> Thanks for the PR, this is a very useful feature but it's also very tricky to get it right.
> 
> As this changes the semantic of the path field in the authorization policy (beta API), I think we should get approval at least from the security work group. Could you discuss it in work group meeting and probably write a quick doc about the change for collecting feedback and helping discussions?

**Comment by mstrYoda** [comment]
> Are you able to drop by a working group meeting to discuss?
> 
> See [#35641 (review)](https://github.com/istio/istio/pull/35641#pullrequestreview-783642732):
> 
> > Thanks for the PR, this is a very useful feature but it's also very tricky to get it right.
> > As this changes the semantic of the path field in the authorization policy (beta API), I think we should get approval at least from the security work group. Could you discuss it in work group meeting and probably write a quick doc about the change for collecting feedback and helping discussions?

Of course, I am willing to attend to wg meeting 🙏

**Comment by mstrYoda** [comment]
> Are you able to drop by a working group meeting to discuss?
> 
> See [#35641 (review)](https://github.com/istio/istio/pull/35641#pullrequestreview-783642732):
> 
> > Thanks for the PR, this is a very useful feature but it's also very tricky to get it right.
> > As this changes the semantic of the path field in the authorization policy (beta API), I think we should get approval at least from the security work group. Could you discuss it in work group meeting and probably write a quick doc about the change for collecting feedback and helping discussions?

Hi @craigbox, can we reopen this PR? This is a highly wanted feature from community.

I try to attend one of the sec. workgroup meeting 🙏

**Comment by nberkley-gp** [comment]
Is there any chance of getting this merged? The lack of unanchored string matches in auth policies is an issue for multiple use cases and all the related PRs/Issues seem to have gone silent for a while.

**Comment by craigbox** [comment]
Did @mstrYoda meet with the working groups?
@howardjohn for further comment

**Comment by mstrYoda** [comment]
> Did @mstrYoda meet with the working groups?
> 
> @howardjohn for further comment

I did not after I saw that there will be a proposal doc. But I can attend to the closest meeting ASAP if you want to 👍

**Comment by howardjohn** [comment]
This PR cannot be accepted as is (not the feature in general, just this PR from a process point of view); this will require an API change - even if not changing the proto, the docs at least.

To keep the discussion focused lets consolidate on https://github.com/istio/api/pull/2173 or the issue https://github.com/istio/istio/issues/16585

Thanks!

**Pull request closed without merging**